### PR TITLE
Pico SDK - Fix Compiler Definitions & RISC-V RP2350 support

### PR DESCRIPTION
**Pull request by iCMDdev**
This is a PR for fixing the issues outlined in #59.

### CODE REVIEW

**Comment by iCMDdev** [comment]
As I previously mentioned, this PR addresses #59. 

The Pico SDK examples resulted in build errors with the new 2.0.0 SDK due to the required compiler flags not being passed to `swiftc`. The proposed solution recursively gathers compile definitions for each target, quite similar to [this example from the Raspberry Pi pico-playground repo](https://github.com/raspberrypi/pico-playground/blob/9930bb057b7272176fbcd924b1d5856dfbc07503/standalone/static_sdk/hacky_cmake_helper.cmake).

This PR also addresses issues when building for RP2350's RISC-V mode, though it's currently pending an SDK update. With Pico SDK version 2.0.1, issues (raspberrypi/pico-sdk#1922) will be fixed. One can pre-test the new version by switching to the SDK `develop` branch.

File: pico-blink-sdk/CMakeLists.txt [hunk 1]
```diff
@@ -12,22 +12,85 @@ endif()
 
 
 set(SWIFT_TARGET "armv6m-none-none-eabi") # default for rp2040
+list(APPEND CLANG_ARCH_ABI_FLAGS "-Xcc") 
 
 if(PICO_PLATFORM STREQUAL "rp2350-arm-s")
     message(STATUS "PICO_PLATFORM is set to rp2350-arm-s, using armv7em")
     set(SWIFT_TARGET "armv7em-none-none-eabi")
+    list(APPEND CLANG_ARCH_ABI_FLAGS "-mfloat-abi=soft")
```
[the remark this answers is not in the document]
Comment (iCMDdev) [reply pico-blink-sdk/CMakeLists.txt]: Sure thing! If you have any other suggestions (including coding style / comment-related), please let me know.

**Comment by iCMDdev** [comment]
To be honest, I didn't know that much CMake either - I learned while writing this. Feedback is definitely welcome 😅

File: pico-blink-sdk/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (etcwilde) [reply pico-blink-sdk/CMakeLists.txt]: So I guess the question is whether this ever gets passed into a raw clang invocation or not. Unfortunately the command calling Swift is a custom command, so we can't use the `$<COMPILE_LANGUAGE:Swift>` [generator expression](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:COMPILE_LANGUAGE) or I'd say just use that to add the `-Xcc` here.

If it doesn't get passed into clang, then just keep the `-Xcc` part.

File: pico-blink-sdk/CMakeLists.txt [hunk 2]
```diff
@@ -10,24 +10,85 @@ else()
...
+get_property(COMPILE_DEFINITIONS GLOBAL PROPERTY compilerdefs_list)
+
+# Parse compiler definitions into a format that swiftc can understand
+list(REMOVE_DUPLICATES COMPILE_DEFINITIONS)
+list(PREPEND COMPILE_DEFINITIONS "") # -Xcc -D
+string(REPLACE "$<TARGET_PROPERTY:PICO_TARGET_BINARY_TYPE>" "$<TARGET_PROPERTY:swift-blinky,PICO_TARGET_BINARY_TYPE>" COMPILE_DEFINITIONS "${COMPILE_DEFINITIONS}")
+string(REPLACE ";" " -Xcc -D" COMPILE_DEFINITIONS "${COMPILE_DEFINITIONS}")
+
+# Write the compiler definitions to a file (this way CMake will evaluate the compiler definitions)
+file(GENERATE OUTPUT ${CMAKE_BINARY_DIR}/swiftc_flags.txt CONTENT "${COMPILE_DEFINITIONS}")
```
Comment (etcwilde) [inline pico-blink-sdk/CMakeLists.txt:82]: Shouldn't need this. The `COMMAND` in a custom command expands generator expressions inline.

> Arguments to COMMAND may use [generator expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#manual:cmake-generator-expressions(7)).

https://cmake.org/cmake/help/latest/command/add_custom_command.html

File: pico-w-blink-sdk/CMakeLists.txt [hunk 3]
```diff
@@ -11,11 +11,65 @@ execute_process(COMMAND which swiftc OUTPUT_VARIABLE SWIFTC OUTPUT_STRIP_TRAILIN
...
+get_property(COMPILE_DEFINITIONS GLOBAL PROPERTY compilerdefs_list)
+
+# Parse compiler definitions into a format that swiftc can understand
+list(REMOVE_DUPLICATES COMPILE_DEFINITIONS)
+list(PREPEND COMPILE_DEFINITIONS "") # -Xcc -D
+string(REPLACE "$<TARGET_PROPERTY:PICO_TARGET_BINARY_TYPE>" "$<TARGET_PROPERTY:swift-blinky,PICO_TARGET_BINARY_TYPE>" COMPILE_DEFINITIONS "${COMPILE_DEFINITIONS}")
+string(REPLACE ";" " -Xcc -D" COMPILE_DEFINITIONS "${COMPILE_DEFINITIONS}")
+
+# Write the compiler definitions to a file (this way CMake will evaluate the compiler definitions)
+file(GENERATE OUTPUT ${CMAKE_BINARY_DIR}/swiftc_flags.txt CONTENT "${COMPILE_DEFINITIONS}")
```
Comment (etcwilde) [inline pico-w-blink-sdk/CMakeLists.txt:65]: The `COMMAND` in a custom command expands generator expressions inline.

File: pico-blink-sdk/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (iCMDdev) [reply pico-blink-sdk/CMakeLists.txt]: @etcwilde 
> If it doesn't get passed into clang, then just keep the -Xcc part.

Yeah, this doesn't get passed into clang directly (raw). This is only passed to the custom`swiftc` command to save architecture-specific flags, using `-Xcc`.

Do you think I should rename `CLANG_ARCH_ABI_FLAGS` to something such as `SWIFT_CC_ABI_ARGS` to better reflect this behaivor?

@rauhul 
I refactored the `-Xcc` part in the next commit, can you please check that I understood correctly what you suggested?

**Review by etcwilde — COMMENTED** [review]
Some minor comments.

You might be able to get the `gather_compile_definitions_recursive` down to a single generator expression. I don't love that part because it's something that CMake will already do for you if we weren't using a custom command.

(Probably for someone else, but it might be worth looking into switching the custom command in favor of a normal `add_library` invocation instead and linking things normally)

File: pico-blink-sdk/CMakeLists.txt (hunk 2 above)
Comment (iCMDdev) [reply pico-blink-sdk/CMakeLists.txt]: Right, makes sense. Thanks!

**Comment by iCMDdev** [comment]
> if we weren't using a custom command.

I was wondering why this was used, and I found this [here](https://github.com/swiftlang/swift/blob/9461af4c5af904e5c997ed281778d6a51f1c561a/docs/EmbeddedSwift/IntegratingWithSDKs.md):

> Finally, we need to define the application's build rules in CMake that will be using CMake logic from the Pico SDK. The following content of CMakeLists.txt shows how to manually call swiftc, the Swift compiler instead of using the recently added CMake native support for Swift, so that we can see the full Swift compilation command.

<br>

Regarding a native CMake implementation:

> a normal add_library invocation instead and linking things normally

We should be able to replace the special `swiftc` arguments (such as `-enable-experimental-feature Embedded`, `-target armv6m-none-none-eabi` etc.) with `target_compile_options` calls, right? Is there a limiting factor here? Or, perhaps the idea is to provide an under-the-hood view of the Swift compilation command?

**Comment by rauhul** [comment]
> We should be able to replace the special `swiftc` arguments (such as `-enable-experimental-feature Embedded`, `-target armv6m-none-none-eabi` etc.) with `target_compile_options` calls, right? Is there a limiting factor here? Or, perhaps the idea is to provide an under-the-hood view of the Swift compilation command?

No, if you can make this work with CMake's native Swift support, that would be amazing

**Comment by fabianfett** [comment]
I ran into this trying to get the pico-w blinky example running. It worked perfectly for me. My two cents here: Perfect might be the enemy of good here.

**Comment by iCMDdev** [comment]
> Perfect might be the enemy of good here.

Right, using the CMake native support turned out to be a bit harder than expected. Might still be possible, though.

Perhaps we can and try to address all issues (if there are any remianing ones) with the current apporach, so people can use the new Pico SDK without issues, and work on the CMake native Swift approach in another issue / PR?

**Comment by rauhul** [comment]
> Perhaps we can and try to address all issues (if there are any remianing ones) with the current apporach, so people can use the new Pico SDK without issues, and work on the CMake native Swift approach in another issue / PR?

I think this looks good to merge with hopes of adopting the swift cmake support later.

I'd like to get @kubamracek to review this before merging though.

**Comment by kubamracek** [comment]
Agree with all that's being said, and I'll throw in that we do have a couple of existence proofs of using the CMake native Swift support, e.g. here <https://github.com/apple/swift-embedded-examples/blob/main/esp32-led-blink-sdk/main/CMakeLists.txt>. But yes, let's merge this PR now as is.

**Review by kubamracek — APPROVED** [review]

**Comment by iCMDdev** [comment]
> Agree with all that's being said, and I'll throw in that we do have a couple of existence proofs of using the CMake native Swift support, e.g. here <https://github.com/apple/swift-embedded-examples/blob/main/esp32-led-blink-sdk/main/CMakeLists.txt>. But yes, let's merge this PR now as is.

I'll look into that 👍

**Comment by iCMDdev** [comment]
I'm no

**Comment by mpapple-swift** [comment]
@rauhul Please review changes completed and approve.  Support for RISC-V on the 2350 for an embedded Swift example would be awesome!

**Comment by iCMDdev** [comment]
Small reminder that this still requires a Pico SDK update in order to run on the RISC-V cores (the patch is in the development branch, should release with 2.0.1). 

So if you want to try this yourself, you currently need to use the development (nightly) branch of the Pico SDK.

Maybe I should add this info in the example's README and change it when the new SDK version comes out.

**Comment by mpapple-swift** [comment]
> Small reminder that this still requires a Pico SDK update in order to run on the RISC-V cores (the patch is in the development branch, should release with 2.0.1).
> 
> So if you want to try this yourself, you currently need to use the development (nightly) branch of the Pico SDK.
> 
> Maybe I should add this info in the example's README and update it when the new SDK version comes out.

@iCMDdev agreed - this would save developers some headaches.  Any idea when the Pico SDK 2.0.1 will be released?

**Comment by iCMDdev** [comment]
> Any idea when the Pico SDK 2.0.1 will be released?

I'm not sure...

> agreed - this would save developers some headaches

Done, what do you think?

**Review by mpapple-swift — COMMENTED** [review]
LGTM - thanks

**Comment by iCMDdev** [comment]
Thanks for merging! 🎉

**Pull request merged**